x/stake/keeper.go

```diff
@@ -33,319 +33,315 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
-		addr := AddrFromKey(iterator.Key())
+	// add the actual validator power sorted store
+	maxValidators := k.GetParams(ctx).MaxValidators
+	validators = make([]Validator, maxValidators)
 
-		// iterator.Value is the validator object
-		store.Set(GetToKickOutValidatorKey(addr), iterator.Value())
-		store.Delete(iterator.Key())
+	iterator := store.SubspaceIterator(ValidatorsBondedKey)
+	i := 0
+	for ; iterator.Valid(); iterator.Next() {
+		bz := iterator.Value()
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		validators[i] = validator
+		i++
 	}
 	iterator.Close()
+	return validators[:i] // trim
+}
 
-	// add the actual validator power sorted store
+// get the group of bonded validators sorted by power-rank
+func (k Keeper) GetValidatorsBondedByPower(ctx sdk.Context) []Validator {
+	store := ctx.KVStore(k.storeKey)
 	maxValidators := k.GetParams(ctx).MaxValidators
-	iterator = store.ReverseIterator(subspace(ValidatorsKey)) // largest to smallest
-	validators = make([]Validator, maxValidators)
+	validators := make([]Validator, maxValidators)
+	iterator := store.ReverseSubspaceIterator(ValidatorsByPowerKey) // largest to smallest
 	i := 0
-	for ; ; i++ {
+	for {
 		if !iterator.Valid() || i > int(maxValidators-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
 		var validator Validator
-		err := k.cdc.UnmarshalBinary(bz, &validator)
-		if err != nil {
-			panic(err)
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		if validator.Status == sdk.Bonded {
+			validators[i] = validator
+			i++
 		}
-		validators[i] = validator
-
-		// remove from ToKickOut group
-		store.Delete(GetToKickOutValidatorKey(validator.Address))
-
-		// also add to the recent validators group
-		store.Set(GetRecentValidatorKey(validator.Address), bz)
-
 		iterator.Next()
 	}
+	return validators[:i] // trim
+}
 
-	// add any kicked out validators to the acc change
-	iterator = store.Iterator(subspace(ToKickOutValidatorsKey))
+// XXX TODO build in consideration for revoked
+//
+// Update the validator group and kick out any old validators. In addition this
+// function adds (or doesn't add) a validator which has updated its bonded
+// tokens to the validator group. -> this validator is specified through the
+// updatedValidatorAddr term.
+//
+// The correct subset is retrieved by iterating through an index of the
+// validators sorted by power, stored using the ValidatorsByPowerKey. Simultaniously
+// the current validator records are updated in store with the
+// ValidatorsBondedKey. This store is used to determine if a validator is a
+// validator without needing to iterate over the subspace as we do in
+// GetValidators.
+func (k Keeper) updateValidators(ctx sdk.Context, store sdk.KVStore, updatedValidatorAddr sdk.Address) (updatedIsBonded bool) {
+	updatedIsBonded = false
+
+	// clear the current validators store, add to the ToKickOut temp store
+	toKickOut := make(map[string][]byte) // map[key]value
```

Comment:
Some concern with using native maps (non-deterministic), see https://github.com/cosmos/cosmos-sdk/issues/553

I think map order-of-iteration shouldn't matter in this function though?

Reply:
This doesn't matter here, all the elements in this map are processed, and there is no situation with writes which depend on the order of this map whatsoever within this function